Project/Source/Components/UI/ComponentVideo.cpp

```diff
@@ -0,0 +1,391 @@
+#include "ComponentVideo.h"
+
+#include "Application.h"
+#include "GameObject.h"
+#include "Modules/ModuleUserInterface.h"
+#include "Modules/ModulePrograms.h"
+#include "Modules/ModuleCamera.h"
+#include "Modules/ModuleRender.h"
+#include "Modules/ModuleTime.h"
+#include "Components/ComponentAudioSource.h"
+#include "Components/UI/ComponentTransform2D.h"
+#include "Resources/ResourceVideo.h"
+#include "Utils/ImGuiUtils.h"
+#include "Utils/Logging.h"
+
+#include "Math/TransformOps.h"
+
+extern "C" {
+#include "libavformat/avformat.h"
+#include "libswscale/swscale.h"
+}
+
+#include "Utils/Leaks.h"
+
+#define JSON_TAG_VIDEOID "VideoId"
+#define JSON_TAG_IS_FLIPPED "VerticalFlip"
+
+char av_error[AV_ERROR_MAX_STRING_SIZE] = {0};
+#define av_err2str(errnum) av_make_error_string(av_error, AV_ERROR_MAX_STRING_SIZE, errnum)
+
+ComponentVideo::~ComponentVideo() {
+	CloseVideoReader();
+
+	// Release GL texture
+	glDeleteTextures(1, &frameTexture);
+}
+
+void ComponentVideo::Init() {
+	// Load shader
+	imageUIProgram = App->programs->imageUI;
+
+	// Set GL texture buffer
+	glGenTextures(1, &frameTexture);
+	glBindTexture(GL_TEXTURE_2D, frameTexture);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+
+	if (videoID) OpenVideoReader(App->resources->GetResource<ResourceVideo>(videoID)->GetAssetFilePath().c_str());
+}
+
+void ComponentVideo::Update() {
+	if (videoID != 0) {
+		elapsedVideoTime += App->time->GetDeltaTime();
+		if (elapsedVideoTime > videoFrameTime || videoFrameTime == 0) { // TODO: this will give an offset of 1 frame between video and audio!
+			ReadVideoFrame();
+		}
+
+		if (elapsedVideoTime > audioFrameTime) {
+			ReadAudioFrame(); //each packet contains complete frames, or multiple frames in the case of audio.
+		}
+		// audioPlayer.UpdateStreamData(audioFrameData, audioPlayer.checkFramesSync());
+	}
+}
+
+void ComponentVideo::OnEditorUpdate() {
+	if (ImGui::Checkbox("Active", &active)) {
+		if (GetOwner().IsActive()) {
+			if (active) {
+				Enable();
+			} else {
+				Disable();
+			}
+		}
+	}
+	ImGui::Separator();
+
+	ImGui::ResourceSlot<ResourceVideo>(
+		"Video Resource",
+		&videoID,
+		[this]() { RemoveVideoResource(); },
+		[this]() { OpenVideoReader(App->resources->GetResource<ResourceVideo>(videoID)->GetAssetFilePath().c_str()); });
+
+	std::string removeButton = std::string(ICON_FA_TIMES "##") + "video";
+	if (ImGui::Button(removeButton.c_str())) {
```

Comment:
Care because this was added in other PR in ImGuiUtils, same we have in ResourceSlot<ResourceTexture>. Currently in develop